lib/phobos/actions/process_message_batch.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module Phobos
+  module Actions
+    class BatchMessage
+      attr_accessor :key, :partition, :offset, :payload
+
+      def initialize(key:, partition:, offset:, payload:)
+        @key = key
+        @partition = partition
+        @offset = offset
+        @payload = payload
+      end
+
+      def ==(other)
+        [:key, :partition, :offset, :payload].all? do |s|
+          send(s) == other.send(s)
+        end
+      end
+    end
+
+    class ProcessMessageBatch < ProcessMessage
```

Comment:
You are subclassing ProcessMessage probably to get access to retry/handle error, even though conceptually they are very different. However I feel this is kind of a shortcut and it could be worth investigating for various reasons:
- you get stuff you dont want/need such as `force_encoding`, duplicating a constant `MAX_SLEEP_INTERVAL`
- the code is incrementing instrumentation regarding `listener.process_message` but this is clearly not the case, this is `listener.process_message_batch`
- the code is incrementing instrumentation regarding `listener.retry_handler_error` but in this case I think it should be `listener.retry_handler_error_batch` and the message should mention `error processing batch` instead.
- I would personally move error handling in `process_message.rb` to a module `process_error_handling.rb`, specifically the methods `retry_count`, `increment_retry_count`, `backoff`, `backoff_interval`, `snooze` looks like they can be moved easily. Only `handle_error` needs special attention with its logging and instrumentation. could possibly stay in `process_message.rb` but reimplemented in `process_message_batch.rb` using correct instrumentation and logging.
- include the ProcessErrorHandling and then remove the subclassing
- update the Readme to document the 2 new instrumentation options available

Reply:
So re a separate `process_batch_inline` - I didn't really see the value in a separate instrumentation here because literally all that happens was it went into ProcessBatch and then immediately went into ProcessBatchMessages. That's also why I called it `process_messages` - it's a relic from when it lived inside ProcessBatch.

Based on your comment below where we skip ProcessBatch entirely, this all works quite nicely. :) I'll make the changes.